src/os/bluestore/BlueFS.cc

```diff
@@ -3757,26 +3759,22 @@ ceph::bufferlist BlueFS::FileWriter::flush_buffer(
 {
   ceph_assert(ceph_mutex_is_locked(this->lock) || file->fnode.ino <= 1);
   ceph::bufferlist bl;
-  if (partial) {
-    tail_block.splice(0, tail_block.length(), &bl);
-  }
-  dout(20) << __func__ << " tail is" << std::hex << bl.length() << dendl;
-  ceph_assert(length >= bl.length());
-  const auto remaining_len = length - bl.length();
-  buffer.splice(0, remaining_len, &bl);
+  dout(20) << __func__ << " tail is 0x" << std::hex << remaining_tail
```

Comment:
1) Good idea to get rid of extra bufferlist tail_block.

2) Proposal for different flush algorithm:
```
measure tail_size (payload size in last 4K block)
measure remaining_size (size in last 4K block that is not used)
append_zeros(remaining_size)
if (buffer has more than 4K in current buffer::ptr OR envelope header can still fit) {
  duplicate tail_size bytes 
} else {
  allocate new buffer
  copy tail_size bytes here
}
```
Benefits:
1) we can preallocate `buffer`
2) it can be implemented without adding problematic append_hole() to buffer.h